### PR TITLE
Add some additional E2E tests for client

### PR DESCRIPTION
**Pull request by pkwarren**
In several cases, we were not correctly calling callbacks when an exception occurred while reading a response body, parsing it using a Codec, or attempting to decompress it. Add some additional E2E tests in the new MockWebServerTests to exercise these failures and update the library code to handle the exceptions appropriately.

### CODE REVIEW

File: library/src/main/kotlin/com/connectrpc/impl/ProtocolClient.kt [hunk 1]
```diff
@@ -84,20 +84,31 @@ class ProtocolClient(
...
-                    val responseCodec = serializationStrategy.codec(methodSpec.responseClass)
-                    val responseMessage = responseCodec.deserialize(
-                        finalResponse.message,
-                    )
+                    return@httpClientUnary
+                }
+                val responseCodec = serializationStrategy.codec(methodSpec.responseClass)
+                val responseMessage: Output
+                try {
+                    responseMessage = responseCodec.deserialize(finalResponse.message)
```
Comment (pkwarren) [inline library/src/main/kotlin/com/connectrpc/impl/ProtocolClient.kt:92]: We were missing exception handling around deserializing using the Codec, which lead to not notifying callers of a failure.

File: library/src/main/kotlin/com/connectrpc/protocols/ConnectInterceptor.kt [hunk 2]
```diff
@@ -95,13 +95,33 @@ internal class ConnectInterceptor(
                 val responseHeaders =
                     response.headers.filter { entry -> !entry.key.startsWith("trailer-") }
                 val compressionPool = clientConfig.compressionPool(responseHeaders[CONTENT_ENCODING]?.first())
+                val responseBody = try {
+                    compressionPool?.decompress(response.message.buffer) ?: response.message.buffer
```
Comment (pkwarren) [inline library/src/main/kotlin/com/connectrpc/protocols/ConnectInterceptor.kt:99]: Similarly we weren't handling exceptions properly when decompressing data.

File: library/src/main/kotlin/com/connectrpc/protocols/ConnectInterceptor.kt [hunk 3]
```diff
@@ -95,13 +95,33 @@ internal class ConnectInterceptor(
...
+                        ),
+                        tracingInfo = response.tracingInfo,
+                    )
+                }
+                val message: Buffer
                 val (code, exception) = if (response.code != Code.OK) {
-                    val error = parseConnectUnaryException(code = response.code, response.headers, response.message.buffer)
+                    val error = parseConnectUnaryException(code = response.code, response.headers, responseBody)
+                    // We've already read the response body to parse an error - don't read again.
+                    message = Buffer()
```
Comment (pkwarren) [inline library/src/main/kotlin/com/connectrpc/protocols/ConnectInterceptor.kt:119]: @erawhctim - This fixes the issue before where we'd both read a JSON error response and then read again for the message.

File: okhttp/src/test/kotlin/com/connectrpc/okhttp/MockWebServerTests.kt [hunk 4]
```diff
@@ -45,9 +51,66 @@ class MockWebServerTests {
...
 
-        val host = mockWebServerRule.server.url("/")
+    @Test
+    fun `invalid compressed response data is handled correctly`() = runTest {
+        mockWebServerRule.server.enqueue(
+            MockResponse().apply {
+                addHeader("accept-encoding", "gzip")
+                addHeader("content-encoding", "gzip")
+                setBody("this isn't gzipped")
+                setResponseCode(200)
```
Comment (pkwarren) [inline okhttp/src/test/kotlin/com/connectrpc/okhttp/MockWebServerTests.kt:68]: Tests http status = 200, invalid compressed data.

File: okhttp/src/test/kotlin/com/connectrpc/okhttp/MockWebServerTests.kt [hunk 5]
```diff
@@ -45,9 +51,66 @@ class MockWebServerTests {
...
+        assertThat(response.code).isEqualTo(Code.INTERNAL_ERROR)
+    }
 
+    @Test
+    fun `invalid protobuf response data is handled correctly`() = runTest {
+        mockWebServerRule.server.enqueue(
+            MockResponse().apply {
+                addHeader("accept-encoding", "gzip")
+                addHeader("content-type", "application/proto")
+                setBody("this isn't valid protobuf")
```
Comment (pkwarren) [inline okhttp/src/test/kotlin/com/connectrpc/okhttp/MockWebServerTests.kt:84]: Tests http status = 200, invalid proto message.

File: okhttp/src/test/kotlin/com/connectrpc/okhttp/MockWebServerTests.kt [hunk 6]
```diff
@@ -45,9 +51,66 @@ class MockWebServerTests {
...
+    }
+
+    @Test
+    fun `invalid json response data is handled correctly`() = runTest {
+        mockWebServerRule.server.enqueue(
+            MockResponse().apply {
+                addHeader("accept-encoding", "gzip")
+                addHeader("content-type", "application/json")
+                setBody("{ invalid json")
+                setResponseCode(200)
```
Comment (pkwarren) [inline okhttp/src/test/kotlin/com/connectrpc/okhttp/MockWebServerTests.kt:102]: Tests http status = 200, invalid json message.

**Review by erawhctim — APPROVED** [review]
Looks good from my end; glad those extra error scenarios are under test now.

I was mostly debugging around in `ConnectInterceptor`, and while the `ProtocolClient` changes look reasonable to me, that part could use another set of eyes

**Review by smaye81 — APPROVED** [review]

**Pull request merged**